core/matrix/csr.cpp

```diff
@@ -443,8 +443,26 @@ void Csr<ValueType, IndexType>::read(device_mat_data&& data)
     auto arrays = data.empty_out();
     this->row_ptrs_.resize_and_reset(size[0] + 1);
     this->set_size(size);
-    this->values_ = std::move(arrays.values);
-    this->col_idxs_ = std::move(arrays.col_idxs);
+    // Need to copy instead of move, as move into view data is not supported.
```

Comment:
No, move into view is not supported. It does not make sense to move data from an owning array to a non-owning one. I think the behaviour in array is correct. To have correct behaviour, we can either have all cases just copy, or copy only when the arrays of `this` are non-owning, and I think it is better to avoid deep copies wherever possible, so I think the implementation here is a good solution.

Reply:
I will create that PR.

Reply:
I think the behaviour of `const device_mat_data&` overload does copy. This function is just a helper that reduces code duplication by only taking in a temporary object. 

I agree that it is an edge-case, but it is very useful in implementing read for batched objects. If and when we implement non-uniform batch objects, I think we should follow a similar approach.

Reply:
Sadly the `const device_mat_data&` overload only copies the input, but it does not copy into the matrix's arrays, it moves again. So IMO, making the `const device_mat_data&` overload also copy the arrays would fix the issue here, while keeping the current move semantics.

Reply:
I would suggest that the input itself is not copied, but instead only the arrays are copied. This should have no overhead.

Moving data into views doesn't fail, it does exactly what is should do. The move overload takes ownership of the input data, and thus the matrix takes ownership of the stored arrays. So afterwards the matrix can't have array views anymore. I agree that we should clarify this, but I think this is the correct behavior.

Reply:
I tend to agree, reading into a view was never something we really considered before. This only matters for matrix types that reuse the arrays anyways (Csr and Coo), so it's not a notable maintenance overhead to duplicate a tiny bit of the implementation. We should have tests for both cases in place though

Reply:
I am fine with copying the arrays instead of copying the input. As we probably do that in another PR, maybe we push this PR forward with these changes ?